Logging.md

```diff
@@ -0,0 +1,113 @@
+# JULEP Logging
+
+- **Title:** A unified logging interface
+- **Author:** Chris Foster <chris42f@gmail.com>
+- **Created:** February 2017
+- **Status:** work in progress
+
+## Abstract
+
+Production systems need robust and consistent logging; casual users want logging
+to be as simple as `println()`.  The logging interface in Base should support
+both these needs in a simple, flexible and efficient way.
+
+In particular, this proposal seeks to:
+* Encourage package authors to include logging by keeping usage simple but
+  making it efficient and convenient.
+* Add a minimum number of features and concepts required to support log
+  filtering, formatting, and dispatch in production environments.
+* Unify logging under a common interface to avoid inconsistency of logging
+  configuration and handling between packages.
+
+## Desirable features
+
+The desirable features fit into three rough categories - simplicity, flexibility
+and efficiency.
+
+Logging should be **simple to use** so that package authors can reach for `info`
+rather than `println()`.  We should have:
+
+* A minimum of syntax - ideally just a logger verb and the message in most
+  cases.  Context information for log messages (file name, line number, module,
+  stack trace, etc.) should be automatically gathered without a syntax burden.
+* Freedom in formatting the log message - simple string interpolation,
+  `@sprintf` and `fmt()`, etc should all be fine.
+* Zero logger setup for simple uses.
+* Easy filtering of log messages.
+* Clear guidelines about the meaning and appropriate use of standard log levels
+  for consistency between packages, along with guiding the appropriate use of
+  logging vs stdout.
+* To encourage the use of logging over ad hoc console output, the default
+  console log handler should emphasize the *log message*, and metadata should be
+  printed in a non-intrusive way.
+
+
+The API should be **flexible enough** for advanced users:
+
+* For all packages using the standard logging API, it should be simple to
+  intercept, filter and redirect logs in a unified and centrally controlled way.
+* Log records are more than a string: loggers typically gather context
+  information both lexically (eg, module, file name, line number) and
+  dynamically (eg, time, stack trace, thread id).  The API should preserve this
+  structured information, perhaps as a LogRecord type.
+* Formatting and dispatch of log records should be in the hands of the user if
+  they need it. For example, a log handler library may need to write json
+  records across the network to a log server.
+* It should be possible to log to a user defined log context; automatically
+  choosing a context for zero setup logging may not suit all cases.  For
+  example, in some cases we may want to use a log context explicitly attached to
+  a user-defined data structure.
```

Comment:
This is basically the "dependency injection" case - needs some more thought about how loggers are associated with contexts.